snowpark_style_guide.md

```diff
@@ -0,0 +1,341 @@
+df = df.with_column("sum", df["a"] +  df["b"]).sort(df["a"]).show()
+
+# performant
+df = df.with_columns(
+   ["mean", "sum"],
+   [(df["a"] + df["b"]) / 2, df["a"] + df["b"]]
+).sort(df["a"]).show()
+```
+
+# Expressions
+
+Chaining of expressions are widely used as they represent atomic logic. Here we are give
+some best practice based on our experiences for writing the expressions.
+
+## Multiple-Lines
+
+In general, it should be avoided to lay out all the expression into a single line which makes code
+difficult to read with expressions left behind getting easier to be omitted.
+
+Wrapping the multi-lines expressions within parentheses provides a graceful way to laying out the code.
+
+```python
+# all in one line
+df = df.select('col_a', 'col_b').filter('col_a' == 'value').filter('cob_b' == True).join(another_dataframe, 'col_c').drop('col_d')
+
+# multiple lines wrapped in parentheses
+df = (
+    df
+    .select('col_a', 'col_b')
+    .filter('col_a' == 'value')
+    .filter('cob_b' == True)
+    .join(another_df, 'col_c')
+    .drop('col_d')
+)
+
+```
+
+# Explicitness
+
+## Calling Methods and Functions with Parameter Names
+
+It is a good habit to specify the names of parameters when calling methods or functions which helps understand which
+roles the parameters play instead of constantly referencing to the API doc to figure out.
+
+```python
+# without parameter name
+df = df.join(another_df, 'id', 'inner')
+df = df.unpivot("sales", "month", ["jan", "feb"]).sort("empid")
+
+# with parameter name
+df = df.join(another_df, using_columns='id', how='inner')
+df = df.unpivot(value_column="sales", name_column="month", column_list=["jan", "feb"]).sort("empid")
+```
+
+# User Defined Functions (UDFs)
+
+[User Defined Functions (UDFs)][udf-service-doc] let you extend the system to perform customized operations and Snowpark
+lets you use the Python programming language to manipulate data and return either scalar or tabular results.
+
+UDFs could be registered in several ways and each has its own advantages and limitations.
+
+
+## Registration
+
+**calling `snowflake.snowpark.functions.udf` function**
+
+The `snowflake.snowpark.functions.udf` function will register the function within the most
+recently created session. If there is only one single active session in the program, it will use that session
+for registration. However, if there are multiple active sessions in your
+program and a different session is wanted, the `session` parameter has to be explicitly set.
+In this case, you may need to keep a record of all the sessions and maintain extra logic to decide on which
+session to use.
+
+```python
+from snowflake.snowpark.functions import udf
+
+def add(x, y):
+     return x + y
+
+# without session parameter, the most recently created session will be used
+udf(
+   add,
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()]
+)
+
+# or explicitly specifying an active session
```

Comment:
Should we say that we recommend that `session` should be set anytime

Reply:
emmm, do you mind explain it a bit more? user should always pass a session when registering?

Reply:
yes. That's my thought but might not be correct 🤦

Reply:
added one sentence to recommend it